### ISSUE TITLE
internal builtin: for reimplementing (some) external commands in the shell

### ISSUE THREAD

**Issue by andychu** [oils-for-unix/oils#684]
bash already has `test` / `[`, `kill`, `true`/`false` and `time` (which is a keyword, not a builtin).

We already have

- `readlink` -- to support finding the current file portably, e.g. #587 

Oil has reason for:

- `sleep` -- mentioned for ble in #653 
- `env` -- because to support Oil assignment, we write `env PYTHONPATH=. foo.py` rather than `PYTHONPATH=. foo.py`

Any others?

**Comment by akinomyoga** [comment]
`echo` and `printf` are also utilities supported as both builtins and external commands. I checked such Bash builtins with the following command.

```bash
$ for b in $(enable |tr ' ' '\n' | sort -u); do \which $b &>/dev/null && echo $b; done
```

Most of them seem to be the command names specified in the following section of POSIX (`newgrp` is not supported as a builtin of Bash).

> [*POSIX XCU 2.9.1 / Command Search and Execution / Item 1. d*](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_09_01_01)
>
> d. If the command name matches the name <sup>[XSI] [Option Start]</sup> of the `type` or `ulimit` utility, or <sup>[Option End]</sup> of a utility listed in the following table, that utility shall be invoked.
>
> `alias`, `bg`, `cd`, `command`, `false`, `fc`, `fg`, `getopts`, `hash`, `jobs`, `kill`, `newgrp`, `pwd`, `read`, `true`, `umask`, `unalias`, `wait`

The remaining builtins are `[`, `test`, `echo`, and `printf`.

**Comment by akinomyoga** [comment]
> https://github.com/oilshell/oil/issues/653#issuecomment-606430377
>
> - I don't think Oil will have enable and shared libraries for quite awhile, but I think we can put builtin sleep in somehow.

To run `ble.sh`, the builtin `sleep` doesn't need to be a loadable builtin (like in Bash), but it is fine to make it a normal builtin specially implemented in Oil like `readlink`.

**Comment by andychu** [comment]
Actually right now `readlink` is within the OIl binary, but it's invoked externally like busybox:

```
$ ln -s oil.ovm readlink
$ ./readlink  # now Oil be haves like readlink
```

That's because I didn't have something like #588 .  I don't want to  clobber `readlink` if it's on the system, for compatibility.  But I wanted something on OS X where readlink isn't available.

So whatever solution we come up with -- `sleep` and `readlink` will both use it.  Basically they can be looked up *after* external binaries, not before.  I don't want to hard-code all that logic, hence something general like #588 .

**Comment by akinomyoga** [comment]
> Basically they can be looked up after external binaries, not before.

OK... but `ble.sh` wants to use *a builtin sleep* instead of external command `sleep` because an external command `sleep` does not necessarily support millisecond resolution and also has an overhead larger than milliseconds to spawn an external process. If the extra builtins shouldn't hide external commands, the name of the sleep builtin can be changed from `sleep`.

**Comment by andychu** [comment]
Or we could have an `internal` builtin that's used as a namespace, so you could do either

```
internal ln target src
```

or

```
ln() {
  internal ln "$@"
}
```

**Comment by happysalada** [comment]
I think you are right, as long as there exist a builtin for a particular function, then we could do something in nix to replace the designed functions with the builtins. Whatever forms they have, as long as the builtins are there with some name, we can arrange to test the replacement in nix.
I think the next step on that, is for me to get a baseline of the time it takes to build stdenv. The best might be if there was a way to get a kind of flame graph to find what externals are being called. (just to get a sense of what externals it's really worth having a builtin for).

**Comment by andychu** [comment]
Yeah I have thought about the flame graph problem a lot, which is very related to the tracing threads on Zulip

One problem is tracking process-based concurrency, i.e. fork and wait

The tracer has the start of that I think.  We should talk about it.

I think there was some prior art in profiling Debian's bootup sequence but I don't remember offhand at the moment

**Comment by happysalada** [comment]
We should definitely talk about it. Would you have some time next week to do another one of those dev talks ? I think it would be nice if it was one item on the agenda. The other being just going once through the technical steps of the build process. I am personally fairly available next week and happy to adapt to your availabilities.

**Comment by andychu** [comment]
Sure we could do it next Friday (not tomorrow) at same time, 4pm EST?  I can post a message on Zulip about it

**Comment by happysalada** [comment]
Friday 3rd of December 4 pm EST is good for me!

**Comment by andychu** [comment]
We now have "private builtins" --  `builtin cat` and `builtin rm` and such, so there is a place to add more if desired
